stellargraph/layer/gcn_lstm.py

```diff
@@ -0,0 +1,347 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2018-2020 Data61, CSIRO
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import warnings
+import numpy as np
+import tensorflow as tf
+from tensorflow.keras import backend as K
+from tensorflow.keras import (
+    activations,
+    initializers,
+    constraints,
+    regularizers,
+    Sequential,
+    Model,
+)
+from tensorflow.keras.layers import Input, Layer, Lambda, Dropout, Reshape, LSTM, Dense
+
+from ..mapper import FullBatchGenerator
+from .misc import SqueezedSparseConversion
+from .preprocessing_layer import GraphPreProcessingLayer
+
+
+class GraphConvolution(Layer):
```

Comment:
As far as I can tell, this differs to the existing `gcn.GraphConvolution` layer in four ways:

1. the adjacency matrix is a constant
2. it doesn't squeeze out the batch dimension
3. it does some transposes in the matrix multiplications
4. it doesn't have special handling for `final_layer=True`

Is this accurate? Could you expand on why 1 and 3 in particular are needed?

Reply:
> As far as I can tell, this differs to the existing `gcn.GraphConvolution` layer in four ways:
> 
> 1. the adjacency matrix is a constant

Yes,  it is. There are two pragmatic reasons for that but neither are binding. 
1. The temporal data  means structure is fixed but features are changing. So its from performance perspective better to just load it at initialization.  However, we can make it more general for dynamic graphs and also to make it more aligned with the already existing `graph_convolution` layer. 
2. The graph convolution happens for each timestep of each observation. For example, for a  sequence of 10 timesteps, gcn is performed for each one of them with the same static graph but different time (feature), so seems like a big overhead of passing the adjacency matrix in each build. 
I am sure there is a more optimal way of doing this. This was just meant to produce a working version.

> 2. it doesn't squeeze out the batch dimension

Yes and this is by design. `Keras` require the batch dimension whereas `GCN` doesn't so even though a unit  batch dimension is passed it gets squeezed inside the layer. However, in the `gcn_lstm`, the data is passed as a 3D tensor `[batch, sequence, nodes]`, don't need to squeeze the batch dimension. 
Maybe we can add the not squeezing as an option to the existing `graph_convolution` layer but didn't want to update it as a first pass until I have the actual dimensions I need to `gcn_lstm`  figured out.  This can be merged in the more polished version. 

> 3. it does some transposes in the matrix multiplications

Yes,  due to the dimension. Typically, you have nodes times features dimensions. Here we have batch times features times nodes dimension. Since graph convolution happens on the feature dimension, so i transpose it before convolving on the feature dimension and then flip it back. 

> 4. it doesn't have special handling for `final_layer=True`

Since here `graph_convolution` is stacked over the RNN layer, i.e. it feeds into the RNN so its never a `final_layer`, so its never true. I can pass it as false when I define the architecture, so this can easily be handled with the `graph_convolution` layer we already have. I am just keeping it false without any special handling. I'll add comment to explain this in the code. 

> Is this accurate? Could you expand on why 1 and 3 in particular are needed?

See above. Any thoughts on these design choices are really appreciated!

Reply:
> Yes, it is. There are two pragmatic reasons for that but neither are binding.

Makes sense; seems fine to start with it as written for now 👍 

> However, in the gcn_lstm, the data is passed as a 3D tensor [batch, sequence, nodes]

And `batch > 1` here?

> Yes, due to the dimension. Typically, you have nodes times features dimensions. Here we have batch times features times nodes dimension. Since graph convolution happens on the feature dimension, so i transpose it before convolving on the feature dimension and then flip it back.

Ah, right, our other GCN layers have input `[batch, nodes, features]`, makes sense!

> Since here graph_convolution is stacked over the RNN layer, i.e. it feeds into the RNN so its never a final_layer, so its never true. I can pass it as false when I define the architecture, so this can easily be handled with the graph_convolution layer we already have. I am just keeping it false without any special handling. I'll add comment to explain this in the code.

The other GCN layers are usually stacked into another dense layer to compute predictions (or similar), so they're not the "final layer" of the (whole) model either. The `final_layer` handling is so that the full-batch layer only yields predictions for the nodes that were passed to the `.flow` method, rather than all nodes.

Given that, this code currently seems to just yield predictions for all nodes. I think this means it's not possible to train on only a subset of labelled nodes? (For instance, do a train/test split across the nodes, rather than the times?) I guess this isn't too important for these sort of datasets, because it's time that is interesting and so it's assumed we'll always have information for every node?